File: en/java-spring-demo/src/main/java/com/itangcent/springboot/demo/controller/SecurityCtrl.java
package com.itangcent.springboot.demo.controller;

import com.itangcent.springboot.demo.common.model.TestJsonFieldBean;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.annotation.security.RolesAllowed;

/**
 * Spring Security Authorization Test
 * Demonstrates different authorization annotation usage
 * Including @RolesAllowed and @PreAuthorize annotations
 *
 * @test a
 */
@RestController
@RequestMapping(value = "/test/security")
public class SecurityCtrl {

    /**
     * Custom role authorization test
     * Uses @RolesAllowed annotation for role-based control
     * 
     * @param testJsonFieldBean Test JSON field bean
     * @return Returns the request object as is
     * @roleRequired admin Requires admin role to access
     */
    @RolesAllowed("admin")
    @RequestMapping("/rolesAllowed/custom")
    public TestJsonFieldBean rolesAllowedCustom(
            @RequestBody TestJsonFieldBean testJsonFieldBean) {
        return testJsonFieldBean;
    }

    /**
     * Spring Security expression authorization test
     * Uses @PreAuthorize annotation for authorization control
     * 
     * @param testJsonFieldBean Test JSON field bean
     * @return Returns the request object as is
     * @roleRequired admin Requires admin role to access
     */
    @PreAuthorize("hasRole('admin')")
    @RequestMapping("/rolesAllowed/admin")
    public TestJsonFieldBean rolesAllowedAdmin(
            @RequestBody TestJsonFieldBean testJsonFieldBean) {
        return testJsonFieldBean;
    }

}

File: cn/kotlin-spring-demo/src/main/kotlin/com/itangcent/springboot/kotlin/demo/controller/SecurityCtrl.kt
package com.itangcent.springboot.kotlin.demo.controller

import com.itangcent.common.annotation.RolesAllowedInJava
import com.itangcent.common.model.TestJsonFieldBean
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import javax.annotation.security.RolesAllowed

/**
 * 安全控制器
 * 用于演示Spring Security的角色权限控制功能
 */
@RestController
@RequestMapping("/test/security")
class SecurityCtrl {

    /**
     * 基于标准注解的角色权限控制
     * 演示使用@RolesAllowed注解进行权限控制
     *
     * @param testJsonFieldBean 测试数据对象
     * @return 返回原测试对象
     * @roles admin 需要管理员角色才能访问
     */
    @RolesAllowed("admin")
    @RequestMapping("/rolesAllowed")
    fun rolesAllowed(
            @RequestBody testJsonFieldBean: TestJsonFieldBean): TestJsonFieldBean {
        return testJsonFieldBean
    }

    /**
     * 基于自定义Java注解的角色权限控制
     * 演示使用自定义的@RolesAllowedInJava注解进行权限控制
     *
     * @param testJsonFieldBean 测试数据对象
     * @return 返回原测试对象
     * @roles admin 需要管理员角色才能访问
     */
    @RolesAllowedInJava(value = "admin", open = false)
    @RequestMapping("/rolesAllowed/java")
    fun rolesAllowedJava(
            @RequestBody testJsonFieldBean: TestJsonFieldBean): TestJsonFieldBean {
        return testJsonFieldBean
    }
}
